moshi/src/main/java/com/squareup/moshi/JsonUtf8Reader.kt

```diff
@@ -0,0 +1,1169 @@
+    }
+    peeked = PEEKED_NONE
+    pathIndices[stackSize - 1]++
+    return result
+  }
+
+  @Throws(IOException::class)
+  override fun selectString(options: Options): Int {
+    var p = peeked
+    if (p == PEEKED_NONE) {
+      p = doPeek()
+    }
+    if (p < PEEKED_SINGLE_QUOTED || p > PEEKED_BUFFERED) {
+      return -1
+    }
+    if (p == PEEKED_BUFFERED) {
+      return findString(peekedString, options)
+    }
+    var result = source.select(options.doubleQuoteSuffix)
+    if (result != -1) {
+      peeked = PEEKED_NONE
+      pathIndices[stackSize - 1]++
+      return result
+    }
+    val nextString = nextString()
+    result = findString(nextString, options)
+    if (result == -1) {
+      peeked = PEEKED_BUFFERED
+      peekedString = nextString
+      pathIndices[stackSize - 1]--
+    }
+    return result
+  }
+
+  /**
+   * If `string` is in `options` this consumes it and returns its index. Otherwise this
+   * returns -1 and no string is consumed.
+   */
+  private fun findString(string: String?, options: Options): Int {
+    val i = options.strings.indexOfFirst { it == string }
+    return if (i > -1) {
+      peeked = PEEKED_NONE
+      pathIndices[stackSize - 1]++
+      i
+    } else -1
+  }
+
+  @Throws(IOException::class)
+  override fun nextBoolean(): Boolean {
+    var p = peeked
+    if (p == PEEKED_NONE) {
+      p = doPeek()
+    }
+    return when (p) {
+      PEEKED_TRUE -> {
+        peeked = PEEKED_NONE
+        pathIndices[stackSize - 1]++
+        true
+      }
+      PEEKED_FALSE -> {
+        peeked = PEEKED_NONE
+        pathIndices[stackSize - 1]++
+        false
+      }
+      else -> throw JsonDataException("Expected a boolean but was " + peek() + " at path " + path)
+    }
+  }
+
+  @Throws(IOException::class)
+  override fun <T> nextNull(): T? {
+    var p = peeked
+    if (p == PEEKED_NONE) {
+      p = doPeek()
+    }
+    return if (p == PEEKED_NULL) {
+      peeked = PEEKED_NONE
+      pathIndices[stackSize - 1]++
+      null
+    } else {
+      throw JsonDataException("Expected null but was " + peek() + " at path " + path)
+    }
+  }
+
+  @Throws(IOException::class)
+  override fun nextDouble(): Double {
+    var p = peeked
+    if (p == PEEKED_NONE) {
+      p = doPeek()
+    }
+    if (p == PEEKED_LONG) {
+      peeked = PEEKED_NONE
+      pathIndices[stackSize - 1]++
+      return peekedLong.toDouble()
+    }
+    when {
+      p == PEEKED_NUMBER -> peekedString = buffer.readUtf8(peekedNumberLength.toLong())
+      p == PEEKED_DOUBLE_QUOTED -> peekedString = nextQuotedValue(DOUBLE_QUOTE_OR_SLASH)
+      p == PEEKED_SINGLE_QUOTED -> peekedString = nextQuotedValue(SINGLE_QUOTE_OR_SLASH)
+      p == PEEKED_UNQUOTED -> peekedString = nextUnquotedValue()
+      p != PEEKED_BUFFERED -> throw JsonDataException("Expected a double but was " + peek() + " at path " + path)
+    }
+    peeked = PEEKED_BUFFERED
+    val result: Double = try {
+      peekedString!!.toDouble()
```

Comment:
We can eliminate this `!!` by holding the result of the `when` in a local variable as well as writing it to `peekedString`.